File: BuddyGrammarKit/Sources/BuddyGrammarKit/LanguageSupport.swift
import Foundation

/// Shared normalization for language-scoped text intelligence.
public enum LanguageSupport {
    public static let defaultPrimaryCode = "en"

    /// Returns a canonical primary language code, grouping regional variants
    /// and common ISO-639-3 detector output into the same local model scope.
    public static func primaryCode(for languageCode: String?) -> String {
        let primary = languageCode?
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(of: "_", with: "-")
            .split(separator: "-")
            .first?
            .lowercased() ?? ""
        guard (2...8).contains(primary.count),
              primary.allSatisfy(\.isLetter) else {
            return defaultPrimaryCode
        }
        return iso639ThreeToOne[primary] ?? primary
    }

    public static func usesEnglishPriors(languageCode: String?) -> Bool {
        primaryCode(for: languageCode) == defaultPrimaryCode
    }

    private static let iso639ThreeToOne: [String: String] = [
        "ara": "ar",
        "deu": "de",
        "eng": "en",
        "fra": "fr",
        "hin": "hi",
        "ita": "it",
        "jpn": "ja",
        "kor": "ko",
        "nld": "nl",
        "pol": "pl",
        "por": "pt",
        "rus": "ru",
        "spa": "es",
        "tur": "tr",
        "ukr": "uk",
        "zho": "zh",
    ]
}

File: android/app/src/main/java/com/francescooddo/buddygrammar/core/LanguageSupport.kt
package com.francescooddo.buddygrammar.core

/** Shared language normalization for local text intelligence and personalization. */
object LanguageSupport {
    const val DEFAULT_LANGUAGE_TAG = "en"

    fun preferredTag(editorHintTags: List<String>, deviceTags: List<String>): String =
        (editorHintTags.asSequence() + deviceTags.asSequence())
            .map(String::trim)
            .firstOrNull(String::isNotEmpty)
            ?: DEFAULT_LANGUAGE_TAG

    /** Groups regional variants so learned language habits follow the user across locales. */
    fun scope(languageTag: String?): String {
        val primary = languageTag
            ?.trim()
            ?.replace('_', '-')
            ?.substringBefore('-')
            ?.lowercase()
            .orEmpty()
        val valid = primary.takeIf { it.length in 2..8 && it.all(Char::isLetter) }
            ?: return DEFAULT_LANGUAGE_TAG
        return ISO_639_2_TO_1[valid] ?: valid
    }

    fun usesEnglishPriors(languageTag: String?): Boolean =
        scope(languageTag) == DEFAULT_LANGUAGE_TAG

    private val ISO_639_2_TO_1 = mapOf(
        "ara" to "ar",
        "deu" to "de",
        "eng" to "en",
        "fra" to "fr",
        "hin" to "hi",
        "ita" to "it",
        "jpn" to "ja",
        "kor" to "ko",
        "nld" to "nl",
        "pol" to "pl",
        "por" to "pt",
        "rus" to "ru",
        "spa" to "es",
        "tur" to "tr",
        "ukr" to "uk",
        "zho" to "zh",
    )
}
